alchemist/alchemist-incarnation-biochemistry/src/main/java/it/unibo/alchemist/model/implementations/environments/BioRect2DEnvironmentNoOverlap.java

```diff
@@ -0,0 +1,263 @@
+            final double q1 = yo - m1 * xo;
+            // computes parameter of straight line, perpendicular to the previous, passing through the cell
+            final double m2 = -1 / m1;
+            final double q2 = yn - m2 * xn;
+            // compute intersection between this two straight lines
+            xIntersect = (q2 - q1) / (m1 - m2);
+            yIntersect = m2 * xIntersect + q2;
+        }
+        final Position intersection = new Continuous2DEuclidean(xIntersect, yIntersect);
+        // computes distance between the cell and the first straight line
+        final double cat = intersection.getDistanceTo(possibleObstaclePosition);
+        // if cat is bigger than cellRange, actual cell isn't an obstacle for the cellular movement
+        if (cat >= cellRange) {
+            // so returns an empty optional
+            return Optional.absent();
+        }
+        // otherwise, compute the maximum practicable distance for the cell
+        final double module =  FastMath.sqrt(FastMath.pow((yIntersect - yo), 2) + FastMath.pow((xIntersect - xo), 2));
+        if (module == 0) { // if module == 0 the translation is 0, so return originalPos
+            return Optional.of(originalPos);
+        }
+        // compute the versor relative to requested direction of cell movement
+        final double cat2 = FastMath.sqrt((FastMath.pow(cellRange, 2) - FastMath.pow(cat, 2)));
+        final double distToSum  = originalPos.getDistanceTo(intersection) - cat2;
+        final Position versor = new Continuous2DEuclidean((xIntersect - xo) / module, (yIntersect - yo) / module);
+        // computes vector representing the practicable movement
+        final Position vectorToSum = new Continuous2DEuclidean(distToSum * (versor.getCoordinate(0)), distToSum * (versor.getCoordinate(1)));
+        // returns the right position of the cell
+        final Position result = originalPos.sum(vectorToSum);
+        return Optional.of(result);
+    }
+
+    @Override
+    protected void nodeAdded(final Node<Double> node, final Position position, final Neighborhood<Double> neighborhood) {
+        super.nodeAdded(node, position, neighborhood);
+        if (node instanceof CellWithCircularArea) {
+            final CellWithCircularArea cell = (CellWithCircularArea) node;
+            if (cell.getDiameter() > getMaxDiameterAmongCells()) {
+                biggestCell = Optional.of(cell); 
+            }
+        }
+        if (node instanceof CircularDeformableCell) {
+            final CircularDeformableCell cell = (CircularDeformableCell) node;
+            if (cell.getMaxDiameter() > getMaxDiameterAmongDeformableCells()) {
+                biggestDeformableCell = Optional.of(cell); 
+            }
+        }
+    }
+
+    @Override
+    protected void nodeRemoved(final Node<Double> node, final Neighborhood<Double> neighborhood) {
+        if (biggestCell.isPresent() && biggestCell.get().equals(node)) {
```

Comment:
Refactor this cut/paste, move the condition check before the cast and act accordingly.
Large copies of pieces of code are **bad**.